### ISSUE TITLE
Cholseky promotes Float16 to Float32

### ISSUE THREAD

**Issue by ctkelley** [JuliaLang/julia#44272]
I found this anomaly and the discussion on

https://discourse.julialang.org/t/cholseky-promotes-float16/76772

with @artkuo seems to conclude that it is a bug. 
--------------
I'm seeing Cholesky promote half precision to single on its own. Is this
supposed to happen? lu does not do this.

```
julia> R=rand(Float16,3,3)
3×3 Matrix{Float16}:
 4.00391e-01  8.24219e-01  2.50000e-01
 2.34375e-02  2.90527e-01  1.09375e-01
 7.85156e-01  9.42383e-01  3.76465e-01

julia> P=R'*R
3×3 Matrix{Float16}:
 7.77344e-01  1.07617e+00  3.98438e-01
 1.07617e+00  1.65234e+00  5.92773e-01
 3.98438e-01  5.92773e-01  2.16187e-:q

julia> PF=lu(P)
LU{Float16, Matrix{Float16}}
L factor:
3×3 Matrix{Float16}:
 1.00000e+00  0.00000e+00  0.00000e+00
 7.22168e-01  1.00000e+00  0.00000e+00
 3.70117e-01  1.58325e-01  1.00000e+00
U factor:
3×3 Matrix{Float16}:
 1.07617e+00   1.65234e+00   5.92773e-01
 0.00000e+00  -1.17188e-01  -2.95410e-02
 0.00000e+00   0.00000e+00   1.50299e-03

julia> QF=cholesky(P)
Cholesky{Float32, Matrix{Float32}}
U factor:
3×3 UpperTriangular{Float32, Matrix{Float32}}:
 8.81671e-01  1.22060e+00  4.51912e-01
      ⋅       4.03073e-01  1.02135e-01
      ⋅            ⋅       3.91250e-02
```

**Comment by RalphAS** [comment]
Note that it is just because of the copy logic: `cholesky!()` does not promote:
```
julia-1.6> cholesky!(P)
Cholesky{Float16, Matrix{Float16}}
U factor:
4×4 UpperTriangular{Float16, Matrix{Float16}}:
 0.949  0.977  0.53    0.655
  ⋅     0.908  0.4631  0.5044
  ⋅      ⋅     0.275   0.4866
  ⋅      ⋅      ⋅      0.3481
```

IIUC Git blame ascribes
```
# make a copy that allow inplace Cholesky factorization
@inline choltype(A) = promote_type(typeof(sqrt(oneunit(eltype(A)))), Float32)
@inline cholcopy(A) = copy_oftype(A, choltype(A))
```
to @andreasnoack Does he recall the reasoning?

**Comment by dkarrasch** [comment]
I'm closing this, since it is fixed on master in #41352.
